File: Solutions/cpp/parkinglot/ParkingSpot.cc
#include "ParkingSpot.h"
#include <iostream>



ParkingSpot::ParkingSpot(int spotNumber, SpotType type): spotNumber(spotNumber), type(type), available(true) {}


int ParkingSpot::getSpotNumber() const {
    return spotNumber;
}

SpotType ParkingSpot::getType() const {
    return type;
}

Vehicle* ParkingSpot::getVehicle() const {
    return vehicle;
}

bool ParkingSpot::isAvailable() const {
    return available;
}


bool ParkingSpot::canFitVehicle(const Vehicle* vehicle) const {
    if (vehicle == nullptr) return false;

    switch (vehicle->getType()) {
        case VehicleType::MOTORCYCLE:
            return true; // can park in any spot
        case VehicleType::CAR:
            return type != SpotType::COMPACT;
        case VehicleType::TRUCK:
        case VehicleType::BUS:
            return type == SpotType::LARGE;
    }

    return false;
}


bool ParkingSpot::parkVehicle(Vehicle* vehicle) {
    if (vehicle == nullptr || !canFitVehicle(vehicle)) return false;

    this->vehicle = vehicle;
    this->available = false;
    return true;
}


Vehicle* ParkingSpot::removeVehicle() {
    if (vehicle == nullptr) return nullptr;

    Vehicle* tmp = vehicle;
    vehicle = nullptr;
    available = true;
    return tmp;
}



void ParkingSpot::displayInfo() const {
    std::cout << "Spot " << spotNumber << " (";
    switch (type) {
        case SpotType::COMPACT: std::cout << "Compact"; break;
        case SpotType::REGULAR: std::cout << "Regular"; break;
        case SpotType::LARGE: std::cout << "Large"; break;
    }
    std::cout << "): " << (available ? "Available" : "Occupied");
    if (vehicle) {
        std::cout << " by ";
        vehicle->displayInfo();
    } else {
        std::cout << std::endl;
    }
}

File: Solutions/golang/parkinglot/parking_spot.go
package parkinglot

type ParkingSpot struct {
	spotNumber    int
	vehicleType   VehicleType
	parkedVehicle Vehicle
}

func NewParkingSpot(spotNumber int, vehicleType VehicleType) *ParkingSpot {
	return &ParkingSpot{
		spotNumber:  spotNumber,
		vehicleType: vehicleType,
	}
}

func (s *ParkingSpot) IsAvailable() bool {
	return s.parkedVehicle == nil
}

func (s *ParkingSpot) ParkVehicle(vehicle Vehicle) bool {
	if !s.IsAvailable() {
		return false
	}
	
	// 规则：
	// 1. 摩托车可以停在任何类型的停车位
	// 2. 汽车只能停在汽车停车位
	// 3. 卡车只能停在卡车停车位
	
	vType := vehicle.GetType()
	switch vType {
	case MOTORCYCLE:
		// 摩托车可以停在任何地方
		s.parkedVehicle = vehicle
		return true
	case CAR:
		// 汽车只能停在汽车或更大的车位
		if s.vehicleType == CAR || s.vehicleType == TRUCK {
			s.parkedVehicle = vehicle
			return true
		}
	case TRUCK:
		// 卡车只能停在卡车车位
		if s.vehicleType == TRUCK {
			s.parkedVehicle = vehicle
			return true
		}
	}
	
	return false
}

func (s *ParkingSpot) UnparkVehicle() bool {
	if s.IsAvailable() {
		return false
	}
	s.parkedVehicle = nil
	return true
}

func (s *ParkingSpot) GetSpotNumber() int {
	return s.spotNumber
}

func (s *ParkingSpot) GetVehicleType() VehicleType {
	return s.vehicleType
}

func (s *ParkingSpot) GetParkedVehicle() Vehicle {
	return s.parkedVehicle
}
